src/OpenLoco/Paint/Paint.cpp

```diff
@@ -63,20 +65,107 @@ namespace OpenLoco::Paint
     // 0x0048DDE4
     [[nodiscard]] InteractionArg PaintSession::getStationNameInteractionInfo(const uint32_t flags)
     {
-        _getMapCoordinatesFromPosFlags = flags;
-        registers regs;
-        regs.edi = reinterpret_cast<uint32_t>(*_dpi);
-        call(0x0048DDE4, regs);
-        return InteractionArg{ _sessionInteractionInfoX, _sessionInteractionInfoY, { _sessionInteractionInfoValue }, _sessionInteractionInfoType, _sessionInteractionInfoBh };
+        InteractionArg interaction{};
+
+        // -2 as there are two interaction items that you can't filter out adjust in future
+        if (flags & (1 << (static_cast<uint32_t>(InteractionItem::station) - 2)))
+        {
+            return interaction;
+        }
+
+        auto zoom = (*_dpi)->zoom_level;
+        auto left = (*_dpi)->x >> zoom;
+        auto top = (*_dpi)->y >> zoom;
+        auto right = ((*_dpi)->width >> zoom) + left;
+        auto bottom = ((*_dpi)->height >> zoom) + top;
+
+        for (auto& station : StationManager::stations())
+        {
+            if (station.empty())
+            {
+                continue;
+            }
+
+            if (station.flags & station_flags::flag_5)
+            {
+                continue;
+            }
+
+            if (top > station.label_bottom[zoom])
+            {
+                continue;
+            }
+            if (bottom < station.label_top[zoom])
+            {
+                continue;
+            }
+            if (left > station.label_right[zoom])
+            {
+                continue;
+            }
+            if (right < station.label_left[zoom])
+            {
+                continue;
+            }
```

Comment:
You could create / use some sort of rect struct and then just have a `contains` call. Help reduce the size of the function.